node/rpc/src/mpool_api.rs

```diff
@@ -0,0 +1,176 @@
+pub(crate) async fn estimate_gas_price<DB, KS, MP>(
+    data: Data<RpcState<DB, KS, MP>>,
+    Params(params): Params<(u64, String, u64, TipsetKeys)>,
+) -> Result<String, JsonRpcError>
+where
+    DB: BlockStore + Send + Sync + 'static,
+    KS: KeyStore + Send + Sync + 'static,
+    MP: Provider + Send + Sync + 'static,
+{
+    let (nblocks, sender_str, gas_limit, tsk) = params;
+    let sender = Address::from_str(&sender_str)?;
+    let price = data
+        .mpool
+        .estimate_gas_price(nblocks, sender, gas_limit, tsk)?;
+    Ok(price.to_string())
+}
+
+/// get the sequence of given address in mpool
+pub(crate) async fn get_sequence<DB, KS, MP>(
+    data: Data<RpcState<DB, KS, MP>>,
+    Params(params): Params<(String,)>,
+) -> Result<u64, JsonRpcError>
+where
+    DB: BlockStore + Send + Sync + 'static,
+    KS: KeyStore + Send + Sync + 'static,
+    MP: Provider + Send + Sync + 'static,
+{
+    let (addr_str,) = params;
+    let address = Address::from_str(&addr_str)?;
+    let sequence = data.mpool.get_sequence(&address).await?;
+    Ok(sequence)
+}
+
+/// Return Vec of pending messages in mpool
+pub(crate) async fn pending<DB, KS, MP>(
+    data: Data<RpcState<DB, KS, MP>>,
+    Params(params): Params<(TipsetKeys,)>,
+) -> Result<Pending, JsonRpcError>
+where
+    DB: BlockStore + Send + Sync + 'static,
+    KS: KeyStore + Send + Sync + 'static,
+    MP: Provider + Send + Sync + 'static,
+{
+    let (tsk,) = params;
+    let mut ts = chain::tipset_from_keys(data.store.as_ref(), &tsk)?;
+
+    let (mut pending, mpts) = data.mpool.pending().await?;
+
+    let mut have_cids = HashSet::new();
+    for item in pending.iter() {
+        have_cids.insert(item.cid()?);
+    }
+
+    if mpts.epoch() > ts.epoch() {
+        return Ok(Pending { msgs: pending });
+    }
+
+    loop {
+        if mpts.epoch() == ts.epoch() {
+            if mpts == ts {
+                return Ok(Pending { msgs: pending });
+            }
+
+            // mpts has different blocks than ts
+            let have = data.mpool.as_ref().messages_for_blocks(ts.blocks()).await?;
+
+            for sm in have {
+                have_cids.insert(sm.cid()?);
+            }
+        }
+
+        let msgs = data.mpool.as_ref().messages_for_blocks(ts.blocks()).await?;
+
+        for m in msgs {
+            if have_cids.contains(&m.cid()?) {
+                continue;
+            }
+
+            have_cids.insert(m.cid()?);
+            pending.push(m);
+        }
+
+        if mpts.epoch() >= ts.epoch() {
+            return Ok(Pending { msgs: pending });
+        }
+
+        let mut headers: Vec<BlockHeader> = Vec::new();
+        let parents = ts.parents().cids();
+        for cid in parents {
+            let block: BlockHeader = data
+                .store
+                .as_ref()
+                .get(cid)?
+                .ok_or("can't find block with cid")?;
+            headers.push(block);
+        }
+        ts = Tipset::new(headers)?;
```

Comment:
why not use the tipset_from_keys here as well? You use it above 

(side note I noticed that function is a bit out of date because it was setup before we had the blockstore trait setup)